node/src/components/sync_leaper/leap_activity.rs

```diff
@@ -0,0 +1,482 @@
+
+    pub(crate) fn leap_start(&self) -> Instant {
+        self.leap_start
+    }
+
+    pub(crate) fn sync_leap_identifier(&self) -> &SyncLeapIdentifier {
+        &self.sync_leap_identifier
+    }
+
+    pub(crate) fn peers(&self) -> &HashMap<NodeId, PeerState> {
+        &self.peers
+    }
+
+    pub(crate) fn peers_mut(&mut self) -> &mut HashMap<NodeId, PeerState> {
+        &mut self.peers
+    }
+
+    /// Registers new leap activity if it wasn't already registered for specified peer.
+    pub(crate) fn register_peer(&mut self, peer: NodeId) -> Option<NodeId> {
+        (!self.peers().contains_key(&peer)).then(|| {
+            self.peers.insert(peer, PeerState::RequestSent);
+            peer
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{
+        collections::{BTreeSet, HashMap},
+        time::Instant,
+    };
+
+    use casper_types::{testing::TestRng, ProtocolVersion};
+    use rand::seq::SliceRandom;
+
+    use crate::{
+        components::sync_leaper::{
+            leap_activity::LeapActivity, tests::make_default_sync_leap, LeapActivityError,
+            LeapState, PeerState,
+        },
+        types::{Block, BlockHash, BlockHeader, NodeId, SyncLeap, SyncLeapIdentifier},
+    };
+
+    fn make_random_block_with_height(rng: &mut TestRng, height: u64) -> Block {
+        Block::random_with_specifics(
+            rng,
+            0.into(),
+            height,
+            ProtocolVersion::default(),
+            false,
+            None,
+        )
+    }
+
+    fn make_sync_leap_with_trusted_block_header(trusted_block_header: BlockHeader) -> SyncLeap {
+        SyncLeap {
+            trusted_ancestor_only: false,
+            trusted_block_header,
+            trusted_ancestor_headers: vec![],
+            signed_block_headers: vec![],
+        }
+    }
+
+    fn assert_peers<I>(expected_peers: I, leap_activity: &LeapActivity)
+    where
+        I: IntoIterator<Item = NodeId>,
+    {
+        let expected_peers: BTreeSet<_> = expected_peers.into_iter().collect();
+        let actual_peers: BTreeSet<_> = leap_activity
+            .peers()
+            .iter()
+            .map(|(node_id, _)| *node_id)
+            .collect();
+        assert_eq!(expected_peers, actual_peers);
+    }
+
+    #[test]
+    fn best_response_with_single_peer() {
+        let mut rng = TestRng::new();
+
+        let sync_leap_identifier = SyncLeapIdentifier::sync_to_tip(BlockHash::random(&mut rng));
+
+        let sync_leap = make_default_sync_leap(&mut rng);
+        let peer_1 = (
+            NodeId::random(&mut rng),
+            PeerState::Fetched(Box::new(sync_leap.clone())),
+        );
+
+        let mut leap_activity = LeapActivity {
+            sync_leap_identifier,
+            peers: [peer_1.clone()].iter().cloned().collect(),
+            leap_start: Instant::now(),
+        };
+
+        let (actual_sync_leap, actual_peers) = leap_activity.best_response().unwrap();
+
+        assert!(!actual_peers.is_empty());
```

Comment:
nit: if we do `actual_peers.get(0).unwrap()` below, there is no need for this assertion.

Reply:
I'd prefer the explicit check, it is more clear to the reader that we actually expect the peers to be empty.